### PR TITLE
Lightmapper is compatible with clustered light mode

### PR DESCRIPTION
**Pull request by mvaligursky**
- Lightmapper can use clustered lights when baking now - it internally allocates clustered lights structure which is updated with lights during baking.

- Clustered lighting has been improved to support flags on the light: affectLightmapped, affectRealtime.

- Renamed light mask constant to match the light properties (those are all private)
Screenshot 2022-01-17 at 17 49 14

### CODE REVIEW

**Comment by Maksims** [comment]
If there is a way to automatically figure out number of lights per cluster and number of clusters when baking for lightmaps, to avoid hitting a limits with too dense/high number of lights while baking lightmaps?

**Comment by mvaligursky** [comment]
Lights are getting baked one at a time as before, so the clustered structure is set to support up to 4 lights are. time (single texture read, minimum supported) .. 

I considered baking more lights at a time (all that could fit their shadows into the atlas), but this has not been done (yet).

File: src/scene/lightmapper/lightmapper.js [hunk 1]
```diff
@@ -447,7 +472,8 @@ class Lightmapper {
                         meshInstance._shaderDefs |= shaderDefs;
                     }
 
-                    meshInstance.mask = MASK_BAKED;
+                    // only lights that affect lightmapped objects are used on this mesh now that is baked
```
Comment (willeastcott) [inline src/scene/lightmapper/lightmapper.js:475]: `now that *it* is baked`?

File: src/scene/constants.js [hunk 2]
```diff
@@ -617,9 +617,10 @@ export const SHADOWUPDATE_REALTIME = 2;
 export const SORTKEY_FORWARD = 0;
 export const SORTKEY_DEPTH = 1;
 
-export const MASK_DYNAMIC = 1;
-export const MASK_BAKED = 2;
-export const MASK_LIGHTMAP = 4;
+// flags used on the mask property of the Light, and also on mask property of the MeshInstance
+export const MASK_AFFECT_DYNAMIC = 1;
+export const MASK_AFFECT_LIGHTMAPPED = 2;
```
Comment (willeastcott) [inline src/scene/constants.js:622]: Is the engine stadnard `ENUM_TYPE_ENUM_VALUE` or `ENUMTYPE_ENUMVALUE`? While I can find some of the former, the latter seems way more prevalent.

File: src/scene/lightmapper/lightmapper.js (hunk 1 above)
Comment (mvaligursky) [reply src/scene/lightmapper/lightmapper.js]: done, thanks

File: src/scene/constants.js (hunk 2 above)
Comment (mvaligursky) [reply src/scene/constants.js:622]: personally I prefer the more readable version with the extra _

Comment (willeastcott) [reply src/scene/constants.js:622]: OK, so are you advocating migrating the engine's enum constants to the more liberal use of underscore?

Comment (mvaligursky) [reply src/scene/constants.js:622]: I guess so .. looking at it again, it seems more than 50% already use underscores for longer constants .. but I'm not sure it's worth adjusting the rest, especially public constants?

Comment (willeastcott) [reply src/scene/constants.js:622]: :shrug: Just pointing out this is definitely an area of inconsistency in the engine right now.

Comment (willeastcott) [reply src/scene/constants.js:622]: TBH, I wish we just used literal strings for most constants.

**Review by willeastcott — APPROVED** [review]

**Pull request merged**